specs/src/rationale/data_square_layout.md

```diff
@@ -0,0 +1,55 @@
+
+The simplest way we can imagine arranging block data is to simply serialize it all in no particular order, split it into fixed-sized shares, then arrange those shares into the `k * k` matrix in row-major order. However, this naive scheme can be improved in a number of ways, described below.
+
+First, we impose some ground rules:
+
+1. Data must be ordered by namespace. This makes queries into a NMT commitment of that data more efficient.
+1. Since non-blob data are not naturally intended for particular namespaces, we assign reserved namespaces for them. A range of namespaces is reserved for this purpose, starting from the lowest possible namespace.
+1. By construction, the above two rules mean that non-blob data always precedes blob data in the row-major matrix, even when considering single rows or columns.
+1. Data with different namespaces must not be in the same share. This might cause a small amount of wasted block space, but makes the NMT easier to reason about in general since leaves are guaranteed to belong to a single namespace.
+
+Transactions can pay fees for a blob to be included in the same block as the transaction itself. However, we do not want serialized transactions to include the entire blob they pay for (which is the case in other blockchains with native execution, e.g. calldata in Ethereum transactions or OP_RETURN data in Bitcoin transactions), otherwise every node that validates the sanctity of the Celestia coin would need to download all blob data. Transactions must therefore only include a commitment to (i.e. some hash of) the blob they pay fees for. If implemented naively (e.g. with a simple hash of the blob, or a simple binary Merkle tree root of the blob), this can lead to a data availability problem, as there are no guarantees that the data behind these commitments is actually part of the block data.
+
+To that end, we impose some additional rules onto _blobs only_: blobs must be placed is a way such that both the transaction sender and the block producer can be held accountable—a necessary property for e.g. fee burning. Accountable in this context means that
+
+1. The transaction sender must pay sufficient fees for blob inclusion.
+1. The block proposer cannot claim that a blob was included when it was not (which implies that a transaction and the blob it pays for must be included in the same block).
+
+Specifically, blobs must begin at a new share, unlike non-blob data which can span multiple shares. We note a nice property from this rule: if the transaction sender knows 1) `k`, the size of the matrix, 2) the starting location of their blob in a row, and 3) the length of the blob (they know this since they are sending the blob), then they can actually compute a sequence of roots to _subtrees in the row NMTs_. More importantly, anyone can compute this, and can compute _the simple Merkle root of these subtree roots_.
```

Comment:
So here are my thoughts. There is an optimization you can make with it here [ADR 009](https://github.com/celestiaorg/celestia-app/blob/main/docs/architecture/adr-009-non-interactive-default-rules-for-reduced-padding.md) This will require changes to the merkle tree on top of the subtreeroots to be a NMT as well. With https://github.com/celestiaorg/celestia-app/issues/1004 You only get this optimization for big squares and smaller messages. This optimization might be negligible with dense blocks, and very rare for them to align so that the internodes in the Merkle Tree are the parents of the subtree roots.  

Another thing that might be possible is when you want to prove fraudulent shares inside the blob for a State Transition Fraud Proof; then you could prove them to the commitment instead of the row-root. 🤔 Could be cheaper but it might not.

Reply:
Why do you want to use a NMT if all the subtree roots belong to the same namespace? What's the gain you're getting

Reply:
Only if we do #1004 do we get the optimization that I analyzed in the link that I showed in ADR 009. Because if it is not an NMT then we don't have a match of H12. The inner node of the merkle tree would not == to the parent of the subtree root.